x-pack/filebeat/input/entityanalytics/provider/okta/okta.go

```diff
@@ -394,18 +407,155 @@ func (p *oktaInput) doFetch(ctx context.Context, state *stateStore, fullSync boo
 	// have a complete set from that timestamp.
 	query = url.Values{}
 	query.Add("search", fmt.Sprintf(`lastUpdated ge "%s" and status pr`, lastUpdated.Format(okta.ISO8601)))
-	state.next = query.Encode()
+	state.nextUsers = query.Encode()
 
 	p.logger.Debugf("received %d users from API", len(users))
 	return users, nil
 }
 
-func grow(u []*User, n int) []*User {
-	if len(u)+n <= cap(u) {
-		return u
+// doFetchDevices handles fetching device and associated user identities from Okta.
+// If fullSync is true, then any existing deltaLink will be ignored, forcing a full
+// synchronization from Okta.
+// Returns a set of modified devices by ID.
+func (p *oktaInput) doFetchDevices(ctx context.Context, state *stateStore, fullSync bool) ([]*Device, error) {
+	if !p.cfg.WantDevices {
+		return nil, nil
+	}
+
+	var (
+		deviceQuery   url.Values
+		userQueryInit url.Values
+		err           error
+	)
+
+	// Get user changes.
+	if !fullSync && state.nextDevices != "" {
+		deviceQuery, err = url.ParseQuery(state.nextDevices)
+		if err != nil {
+			p.logger.Warnf("failed to parse next query: %v", err)
+		}
+	}
+	if deviceQuery == nil {
+		// Use "search" because of recommendation on Okta dev documentation:
+		// https://developer.okta.com/docs/reference/user-query/.
+		// Search term of "status pr" is required so that we get DEPROVISIONED
+		// users; a nil query is more efficient, but excludes these users.
+		// There is no equivalent documentation for devices, so we assume the
+		// behaviour is the same.
+		deviceQuery = url.Values{"search": []string{"status pr"}}
+	}
+	// Start user queries from the same time point. This must not
+	// be mutated since we may perform multiple batched gets over
+	// multiple devices.
+	userQueryInit = cloneURLValues(deviceQuery)
+
+	var (
+		devices     []*Device
+		lastUpdated time.Time
+	)
+	for {
+		batch, h, err := okta.GetDeviceDetails(ctx, p.client, p.cfg.OktaDomain, p.cfg.OktaToken, "", deviceQuery, p.lim, p.cfg.LimitWindow)
+		if err != nil {
+			p.logger.Debugf("received %d devices from API", len(devices))
+			return nil, err
+		}
+		p.logger.Debugf("received batch of %d devices from API", len(batch))
+
+		for i, d := range batch {
+			userQuery := cloneURLValues(userQueryInit)
+			for {
+				// TODO: Consider softening the response to errors here. If we fail to get users
+				// from a device, do we want to fail completely? There are arguments in both
+				// directions. We _could_ keep a multierror and return that in the end, which
```

Comment:
I personally feel holes/gaps in data are more tricky to rectify rather than getting a clear error response and knowing what exactly is going on even if its at the expense of stopping the flow. Though certain scenarios may demand continuous flows I'm not sure if this meets that criteria. If this suggestion is satisfactory, we can remove the TODO.